api/desecapi/pdns_change_tracker.py

```diff
@@ -0,0 +1,228 @@
+    `modifications` and `deletions` are guaranteed to be disjoint.
+    - If an item is in the set of deletions while being modified, an exception is raised.
+    - If an item is in the set of modifications while being deleted, it is removed from `rr_set_modifications`.
+    """
+
+    def __init__(self, optimize=True):
+        self.optimize = optimize
+        self._domain_additions = set()
+        self._domain_deletions = set()
+        self._rr_set_additions = {}
+        self._rr_set_modifications = {}
+        self._rr_set_deletions = {}
+
+    def __enter__(self):
+        self._rr_set_modifications = {}
+        self._rr_set_deletions = {}
+        post_save.connect(self._on_rr_post_save, sender=RR)
+        post_delete.connect(self._on_rr_post_delete, sender=RR)
+        post_save.connect(self._on_rr_set_post_save, sender=RRset)
+        post_delete.connect(self._on_rr_set_post_delete, sender=RRset)
+        post_save.connect(self._on_domain_post_save, sender=Domain)
+        post_delete.connect(self._on_domain_post_delete, sender=Domain)
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if exc_val:
+            return  # TODO proper error treatment
+
+        for domain_name in self._domain_deletions:
+            normalized_name = domain_name.rstrip('.') + '.'
+            pdns_id = normalized_name.translate(str.maketrans({'/': '=2F', '_': '=5F'}))
+
+            assert domain_name not in (self._rr_set_additions.keys()
+                                       | self._rr_set_modifications.keys()
+                                       | self._rr_set_deletions.keys())
+            _pdns_delete(pdns_id)
+
+        for domain_name in self._rr_set_additions.keys() | self._domain_additions:
+            normalized_name = domain_name.rstrip('.') + '.'
+            pdns_id = normalized_name.translate(str.maketrans({'/': '=2F', '_': '=5F'}))
+
+            if domain_name in self._domain_additions:
+                salt = '%016x' % random.randrange(16 ** 16)
+                payload = {'name': normalized_name, 'kind': 'MASTER', 'dnssec': True,
+                           'nsec3param': '1 0 127 %s' % salt, 'nameservers': settings.DEFAULT_NS}
+                _pdns_post('/zones', payload)
+
+            additions = self._rr_set_additions.get(domain_name, set())
+            modifications = self._rr_set_modifications.get(domain_name, set())
+            deletions = self._rr_set_deletions.get(domain_name, set())
+
+            assert additions & deletions == set()
+            assert modifications & deletions == set()
+
+            # Due to disjoint guarantees with `deletions`, we have four types of RR sets:
+            # (1) purely added RR sets
+            # (2) purely modified RR sets
+            # (3) added and modified RR sets
+            # (4) purely deleted RR sets
+
+            # We send RR sets to PDNS if one of the following conditions holds:
```

Comment:
I really think we should not be talking of "RR sets" (not only here, but generally). It sounds like what's send to pdns is a set of RRs and nothing more. However, it also has a type, a ttl, and a subname, and from pdns perspective also other attributes (e.g. comment). The object at hand is therefore very different from a set of RRs ("RR set"). I believe this is highly confusing, especially as this code now also has sets of RR sets.

The same applies to variable and function naming, which now sometimes includes `rr_set` to signify something that has to do with our `RRset` class. I find this confusing for the same reasons. It also leads to the interesting Stilblüte of that in standard Django terminology, `rr_set.rr_set` would denote the set of records in an `RRset` instance.

I consider "RRset" a name and and not a composition of two words. Thus, I'd argue that the snake_case rules don't apply. Can't we just say "RRset" for the class and "rrset" for the object, as we did until now? There are the above complications with snake_casing this (and there may be more, I haven't thought through queryset field lookup expression, for example), and I'm not sure what the benefit is supposed to be.

Reply:
I can later search and replace and adapt the code to your preference